test/common/router/config_impl_test.cc

```diff
@@ -849,13 +852,134 @@ TEST_F(RouterMatcherHashPolicyTest, HashHeaders) {
     Router::RouteConstSharedPtr route = config.route(headers, 0);
-    EXPECT_FALSE(route->routeEntry()->hashPolicy()->generateHash("", headers).valid());
+    EXPECT_FALSE(
+        route->routeEntry()->hashPolicy()->generateHash("", headers, add_cookie_nop_).valid());
   }
   {
     Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo", "GET");
     headers.addCopy("foo_header", "bar");
     Router::RouteConstSharedPtr route = config.route(headers, 0);
-    EXPECT_TRUE(route->routeEntry()->hashPolicy()->generateHash("", headers).valid());
+    EXPECT_TRUE(
+        route->routeEntry()->hashPolicy()->generateHash("", headers, add_cookie_nop_).valid());
+  }
+  {
+    Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/bar", "GET");
+    Router::RouteConstSharedPtr route = config.route(headers, 0);
+    EXPECT_EQ(nullptr, route->routeEntry()->hashPolicy());
+  }
+}
+
+TEST_F(RouterMatcherHashPolicyTest, HashCookieNoTtl) {
+  NiceMock<Runtime::MockLoader> runtime;
+  NiceMock<Upstream::MockClusterManager> cm;
+  route_config_.mutable_virtual_hosts(0)
+      ->mutable_routes(0)
+      ->mutable_route()
+      ->add_hash_policy()
+      ->mutable_cookie()
+      ->set_name("hash");
+  ConfigImpl config(route_config_, runtime, cm, true);
+
+  EXPECT_FALSE(config.usesRuntime());
+
+  {
+    Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo", "GET");
+    Router::RouteConstSharedPtr route = config.route(headers, 0);
+    EXPECT_FALSE(
+        route->routeEntry()->hashPolicy()->generateHash("", headers, add_cookie_nop_).valid());
+  }
+  {
+    Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo", "GET");
+    headers.addCopy("Cookie", "choco=late; su=gar");
+    Router::RouteConstSharedPtr route = config.route(headers, 0);
+    EXPECT_FALSE(
+        route->routeEntry()->hashPolicy()->generateHash("", headers, add_cookie_nop_).valid());
+  }
+  {
+    Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo", "GET");
+    headers.addCopy("Cookie", "choco=late; hash=brown");
+    Router::RouteConstSharedPtr route = config.route(headers, 0);
+    EXPECT_TRUE(
+        route->routeEntry()->hashPolicy()->generateHash("", headers, add_cookie_nop_).valid());
+  }
+  {
+    uint64_t hash_1, hash_2;
+    {
+      Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo", "GET");
+      headers.addCopy("Cookie", "hash=brown");
+      Router::RouteConstSharedPtr route = config.route(headers, 0);
+      hash_1 =
+          route->routeEntry()->hashPolicy()->generateHash("", headers, add_cookie_nop_).value();
+    }
+    {
+      Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo", "GET");
+      headers.addCopy("Cookie", "hash=green");
+      Router::RouteConstSharedPtr route = config.route(headers, 0);
+      hash_2 =
+          route->routeEntry()->hashPolicy()->generateHash("", headers, add_cookie_nop_).value();
+    }
+    EXPECT_NE(hash_1, hash_2);
```

Comment:
I think this could be a separate test since it's going from "is a hash generated or not" to "are generated hashes the same"

If you stick cm_ and runtime_ as member variables of the test fixture and have a utility function which creates and returns cookie policy, you'll be able to minimize code duplication